python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+        item.setEditable(False)
+
+        # keep tabs of which items we are creating
+        item.setData(True, self.IS_SG_MODEL_ROLE)
+
+        # we have not fetched more data for this item yet
+        item.setData(False, self.SG_ITEM_FETCHED_MORE)
+
+        # attach the nav data for access later
+        self._update_item(item, data)
+
+        # allow item customization prior to adding to model
+        self._item_created(item)
+
+        # set up default thumb
+        self._populate_default_thumbnail(item)
+
+        self._populate_item(item, data)
+
+        self._set_tooltip(item, data)
+
+        # run the finalizer (always runs on construction, even via cache)
+        self._finalize_item(item)
+
+        # identify a parent if none supplied. could be found via the
+        # `parent_url` supplied in the data or the root if no parent item
+        # exists.
+        parent = parent or self.item_from_url(data.get("parent_url")) or \
+            self.invisibleRootItem()
+
+        if row is not None:
+            parent.insertRow(row, item)
+        else:
+            # example of using sort/filter proxy model
+            parent.appendRow(item)
+
+        return item
+
+    def __get_data_cache_path(self, cache_seed=None):
+        """
+        Calculates and returns a cache path to use for this instance's query.
+
+        :param cache_seed: Cache seed supplied to the ``__init__`` method.
+
+        :return: The path to use when caching the model data.
+        :rtype: str
+        """
+
+        # hashes to use to generate the cache path
+        params_hash = hashlib.md5()
+        entity_field_hash = hashlib.md5()
+
+        # even though the navigation path provides a nice organizational
+        # structure for caching, it can get long. to avoid MAX_PATH issues on
+        # windows, just hash it
+        params_hash.update(str(self._path))
+
+        # include the schema generation number for clients
+        params_hash.update(str(self._schema_generation))
+
+        # If this value changes over time (like between Qt4 and Qt5), we need to
+        # assume our previous user roles are invalid since Qt might have taken
+        # it over. If role's value is 32, don't add it to the hash so we don't
+        # invalidate PySide/PyQt4 caches.
+        if QtCore.Qt.UserRole != 32:
+            params_hash.update(str(QtCore.Qt.UserRole))
+
+        # include the cache_seed for additional user control over external state
+        params_hash.update(str(cache_seed))
+
+        # iterate through the sorted entity fields to ensure consistent order
+        for (entity_type, fields) in sorted(self._entity_fields.iteritems()):
+            for field in fields:
+                entity_field_hash.update("%s.%s" % (entity_type, field))
+
+        # convert the seed entity field into a path segment.
+        # example: Version.entity => Version/entity
+        seed_entity_field_path = os.path.join(
```

Comment:
They're using the same technique, but I the values, order, and path are all different enough that I'm not sure there's much help a base class could provide. Let me know if you think that's not true.